[CORE-13368] k/server: change alter user scram credentials api logic

IoannisRP opened the pull request:
Implements: [CORE-13368](https://redpandadata.atlassian.net/browse/CORE-13368)

Normal users should not be able to modify superuser scram credentials.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [x] v25.2.x
- [ ] v25.1.x
- [ ] v24.3.x

## Release Notes

### Improvements

* Via the Kafka API, Credentials of superusers are  permitted to be altered or removed only by an authenticated superuser.

[CORE-13368]: https://redpandadata.atlassian.net/browse/CORE-13368?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

IoannisRP committed:
k/server: change alter user scram credentials api logic

Only superusers should be able to modify superuser. When a normal user
tries to modify a superuser, an error is returned.

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--542098ab-4654-474d-acc5-10a2b59c6520-->
<!--TEST RESULTS-->
<!--aa83d170-598e-45b6-8402-cc49f68e35e6-->
<!--TEST RESULTS-->
<!--e66e3df2-74af-41b4-9da6-056230361f9c-->
<!--TEST RESULTS-->
<!--ae506d39-b59c-4e29-872a-9b04e252c533-->
<!--TEST RESULTS-->
<!--47f0a4ab-7682-4c93-94da-784a2a73a14d-->
<!--TEST RESULTS-->
<!--93d5adb0-d268-43c3-a91d-6ea657b155f1-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72541">build#72541</a></summary>


| test_class            | test_method                         | test_arguments   | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                 | test_history                                                                                                                                                            |
|:----------------------|:------------------------------------|:-----------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-----------------------------------------------------------------------------------------------------------------------------------------------------------------------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| PartitionBalancerTest | test_recovery_mode_rebalance_finish | null             | integration | https://buildkite.com/redpanda/redpanda/builds/72541#01995d6c-57e1-4bce-8db1-6b5396025f78 | FLAKY         | 16/21    | upstream reliability is '81.65584415584416'. current run reliability is '76.19047619047619'. drift is 5.46537 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=PartitionBalancerTest&test_method=test_recovery_mode_rebalance_finish |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72589">build#72589</a></summary>


| test_class                | test_method                                       | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                         |
|:--------------------------|:--------------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| MasterTestSuite           | test_chunk_dl_with_random_http_errors             |                                                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/72589#01996142-fe63-44bf-8ad6-d78f52125ff8 | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                                      |
| MasterTestSuite           | test_upload_limit                                 |                                                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/72589#01996142-fe61-478a-a3ef-ff3b59ced36c | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                                      |
| MasterTestSuite           | test_upload_when_reupload_disabled                |                                                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/72589#01996142-fe61-478a-a3ef-ff3b59ced36c | FAIL          | 0/1      |                                                                                                                                                                         |                                                                                                                                                                                      |
| EndToEndCloudTopicsTest   | test_write                                        | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72589#019961ce-8507-43a1-ba5c-5473dc5e8fcd | FLAKY         | 19/21    | upstream reliability is '98.1675392670157'. current run reliability is '90.47619047619048'. drift is 7.69135 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTest&test_method=test_write                                     |
| EndToEndCloudTopicsTxTest | test_write                                        | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72589#019961ce-8508-4763-9ff1-b652f89bf216 | FLAKY         | 12/21    | upstream reliability is '93.69369369369369'. current run reliability is '57.14285714285714'. drift is 36.55084 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTxTest&test_method=test_write                                   |
| ClusterRateQuotaTest      | test_client_group_produce_rate_throttle_mechanism | null                                                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72589#019961ce-84fd-404e-8b8b-ad64fa4e37e9 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_produce_rate_throttle_mechanism |
| NodeWiseRecoveryTest      | test_recovery_local_data_missing                  | {"wait_for_final_manifest_uploads": false}                                                                                                     | integration | https://buildkite.com/redpanda/redpanda/builds/72589#019961ce-8504-4201-ac04-a5df6b232720 | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_recovery_local_data_missing                  |
| RandomNodeOperationsTest  | test_node_operations                              | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72589#019961ce-8508-4763-9ff1-b652f89bf216 | FLAKY         | 20/21    | upstream reliability is '98.23943661971832'. current run reliability is '95.23809523809523'. drift is 3.00134 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations                          |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72600">build#72600</a></summary>


| test_class                           | test_method                                                     | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                       |
|:-------------------------------------|:----------------------------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync                                      | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/72600#01996300-91df-49da-b681-10cb3ab72f8a | FLAKY         | 19/21    | upstream reliability is '91.28787878787878'. current run reliability is '90.47619047619048'. drift is 0.81169 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync                      |
| ClusterRateQuotaTest                 | test_client_response_and_produce_throttle_mechanism             | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72600#01996300-91da-410f-b546-908d125ef059 | FLAKY         | 15/21    | upstream reliability is '98.69451697127938'. current run reliability is '71.42857142857143'. drift is 27.26595 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_and_produce_throttle_mechanism             |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism                         | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72600#019962fe-8345-4b00-a89e-e6b241e016a4 | FLAKY         | 14/21    | upstream reliability is '99.24050632911391'. current run reliability is '66.66666666666666'. drift is 32.57384 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism                         | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72600#01996300-91db-49ba-b9d8-3e3e95cfc254 | FLAKY         | 15/21    | upstream reliability is '97.51243781094527'. current run reliability is '71.42857142857143'. drift is 26.08387 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism_applies_to_next_request | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72600#019962fe-8345-4423-8427-8210e9620630 | FLAKY         | 16/21    | upstream reliability is '100.0'. current run reliability is '76.19047619047619'. drift is 23.80952 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism_applies_to_next_request | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72600#01996300-91de-4974-aec9-6f5e2c2e97f9 | FLAKY         | 17/21    | upstream reliability is '100.0'. current run reliability is '80.95238095238095'. drift is 19.04762 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| NodeWiseRecoveryTest                 | test_recovery_local_data_missing                                | {"wait_for_final_manifest_uploads": false}                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72600#01996300-91df-4540-bfd1-332fa6cd1624 | FLAKY         | 15/21    | upstream reliability is '100.0'. current run reliability is '71.42857142857143'. drift is 28.57143 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_recovery_local_data_missing                                |
| NodeWiseRecoveryTest                 | test_recovery_local_data_missing                                | {"wait_for_final_manifest_uploads": true}                                    | integration | https://buildkite.com/redpanda/redpanda/builds/72600#019962fe-8347-4cc9-9370-d562b3a22110 | FLAKY         | 12/21    | upstream reliability is '96.57701711491443'. current run reliability is '57.14285714285714'. drift is 39.43416 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_recovery_local_data_missing                                |
| RecoveryModeTest                     | test_rolling_restart                                            | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72600#01996300-91de-4974-aec9-6f5e2c2e97f9 | FLAKY         | 18/21    | upstream reliability is '100.0'. current run reliability is '85.71428571428571'. drift is 14.28571 and the allowed drift is set to 50. The test should PASS             | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RecoveryModeTest&test_method=test_rolling_restart                                                |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72658">build#72658</a></summary>


| test_class                         | test_method                                                     | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                       |
|:-----------------------------------|:----------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest            | test_write                                                      | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72658#019971c5-f859-42be-af92-2e38dbe63fc1 | FLAKY         | 11/21    | upstream reliability is '91.44050104384134'. current run reliability is '52.38095238095239'. drift is 39.05955 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTest&test_method=test_write                                                   |
| ClusterRateQuotaTest               | test_client_group_consume_rate_throttle_mechanism               | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72658#019971c1-bbe4-46e6-a16b-50e8bc3285a2 | FLAKY         | 15/21    | upstream reliability is '91.03313840155946'. current run reliability is '71.42857142857143'. drift is 19.60457 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism               |
| ClusterRateQuotaTest               | test_client_response_and_produce_throttle_mechanism             | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72658#019971c1-bbdd-4e6e-895e-093d75f554d0 | FLAKY         | 16/21    | upstream reliability is '89.72602739726028'. current run reliability is '76.19047619047619'. drift is 13.53555 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_and_produce_throttle_mechanism             |
| ClusterRateQuotaTest               | test_client_response_throttle_mechanism                         | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72658#019971c5-f855-4c6c-bb68-b5299ef68bdb | FLAKY         | 16/21    | upstream reliability is '89.19354838709678'. current run reliability is '76.19047619047619'. drift is 13.00307 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest               | test_client_response_throttle_mechanism_applies_to_next_request | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72658#019971c1-bbdf-4ac4-a10c-f5bca74a80b9 | FLAKY         | 9/21     | upstream reliability is '92.64705882352942'. current run reliability is '42.857142857142854'. drift is 49.78992 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| DatalakeE2ETests                   | test_json_schema_unicode                                        | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "query_engine": "trino"}                                                      | integration | https://buildkite.com/redpanda/redpanda/builds/72658#01997224-e135-4e28-bbbd-a423e1306c32 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DatalakeE2ETests&test_method=test_json_schema_unicode                                            |
| SchemaEvolutionE2ETests            | test_old_schema_writer                                          | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "add_column"}   | integration | https://buildkite.com/redpanda/redpanda/builds/72658#019971c5-f858-48bc-867c-552d955019b0 | FLAKY         | 20/21    | upstream reliability is '98.75776397515527'. current run reliability is '95.23809523809523'. drift is 3.51967 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaEvolutionE2ETests&test_method=test_old_schema_writer                                       |
| NodeWiseRecoveryTest               | test_recovery_local_data_missing                                | {"wait_for_final_manifest_uploads": true}                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/72658#019971c5-f853-4621-926e-d86978578124 | FLAKY         | 8/21     | upstream reliability is '94.51553930530164'. current run reliability is '38.095238095238095'. drift is 56.4203 and the allowed drift is set to 50. The test should FAIL  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_recovery_local_data_missing                                |
| NodeWiseRecoveryTest               | test_recovery_local_data_missing                                | {"wait_for_final_manifest_uploads": true}                                                                                              | integration | https://buildkite.com/redpanda/redpanda/builds/72658#01997224-e135-4e28-bbbd-a423e1306c32 | FLAKY         | 11/21    | upstream reliability is '91.53976311336717'. current run reliability is '52.38095238095239'. drift is 39.15881 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=NodeWiseRecoveryTest&test_method=test_recovery_local_data_missing                                |
| RandomNodeOperationsTest           | test_node_operations                                            | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | integration | https://buildkite.com/redpanda/redpanda/builds/72658#01997224-e135-4e28-bbbd-a423e1306c32 | FLAKY         | 19/21    | upstream reliability is '97.82608695652173'. current run reliability is '90.47619047619048'. drift is 7.3499 and the allowed drift is set to 50. The test should PASS    | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RandomNodeOperationsTest&test_method=test_node_operations                                        |
| CreateTopicReplicaDistributionTest | test_topic_aware_distribution                                   | null                                                                                                                                   | integration | https://buildkite.com/redpanda/redpanda/builds/72658#019971c5-f856-42cd-9caf-2a5d6bce724d | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=CreateTopicReplicaDistributionTest&test_method=test_topic_aware_distribution                     |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72720">build#72720</a></summary>


| test_class                           | test_method                                                     | test_arguments                                                               | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  | test_history                                                                                                                                                                                       |
|:-------------------------------------|:----------------------------------------------------------------|:-----------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ShadowLinkConsumeGroupsMirroringTest | test_continuous_group_sync                                      | {"source_cluster_spec": {"cluster_type": "redpanda"}, "with_failures": true} | integration | https://buildkite.com/redpanda/redpanda/builds/72720#01997622-b614-432c-9197-90fba76acef1 | FLAKY         | 17/21    | upstream reliability is '91.12709832134293'. current run reliability is '80.95238095238095'. drift is 10.17472 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ShadowLinkConsumeGroupsMirroringTest&test_method=test_continuous_group_sync                      |
| ClusterRateQuotaTest                 | test_client_group_consume_rate_throttle_mechanism               | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72720#01997622-b617-48a3-b727-9645fff358c7 | FLAKY         | 12/21    | upstream reliability is '86.80057388809182'. current run reliability is '57.14285714285714'. drift is 29.65772 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_group_consume_rate_throttle_mechanism               |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism                         | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72720#01997623-bd2c-491b-8cd8-6ec609d98dec | FLAKY         | 15/21    | upstream reliability is '93.58490566037736'. current run reliability is '71.42857142857143'. drift is 22.15633 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest                 | test_client_response_throttle_mechanism_applies_to_next_request | null                                                                         | integration | https://buildkite.com/redpanda/redpanda/builds/72720#01997623-bd2d-4049-a8f4-946f6bb9f78c | FLAKY         | 15/21    | upstream reliability is '87.66140602582496'. current run reliability is '71.42857142857143'. drift is 16.23283 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/72733">build#72733</a></summary>


| test_class                          | test_method                                                     | test_arguments                                                                                                                        | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   | test_history                                                                                                                                                                                       |
|:------------------------------------|:----------------------------------------------------------------|:--------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| EndToEndCloudTopicsTest             | test_write                                                      | null                                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-779d-4105-be42-bc2e7f3c4b70 | FLAKY         | 11/21    | upstream reliability is '90.80068143100512'. current run reliability is '52.38095238095239'. drift is 38.41973 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=EndToEndCloudTopicsTest&test_method=test_write                                                   |
| ClusterRateQuotaTest                | test_client_response_and_produce_throttle_mechanism             | null                                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-7796-49a3-afe8-566871b0f18f | FLAKY         | 15/21    | upstream reliability is '94.52449567723343'. current run reliability is '71.42857142857143'. drift is 23.09592 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_and_produce_throttle_mechanism             |
| ClusterRateQuotaTest                | test_client_response_throttle_mechanism                         | null                                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-055b-47c7-9272-344d342ff05e | FLAKY         | 11/21    | upstream reliability is '84.9942726231386'. current run reliability is '52.38095238095239'. drift is 32.61332 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism                         |
| ClusterRateQuotaTest                | test_client_response_throttle_mechanism_applies_to_next_request | null                                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-7799-44c8-9030-47f61004ea29 | FLAKY         | 13/21    | upstream reliability is '95.44740973312402'. current run reliability is '61.904761904761905'. drift is 33.54265 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=ClusterRateQuotaTest&test_method=test_client_response_throttle_mechanism_applies_to_next_request |
| SchemaEvolutionE2ETests             | test_old_schema_writer                                          | {"catalog_type": "rest_hadoop", "cloud_storage_type": 1, "produce_mode": "avro", "query_engine": "trino", "test_case": "drop_column"} | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-055f-43ec-8217-8fbb472df0aa | FLAKY         | 20/21    | upstream reliability is '99.74025974025975'. current run reliability is '95.23809523809523'. drift is 4.50216 and the allowed drift is set to 50. The test should PASS   | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=SchemaEvolutionE2ETests&test_method=test_old_schema_writer                                       |
| DisablingPartitionsTest             | test_disable                                                    | null                                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-0560-49cd-aa5a-4ab252b0436c | FLAKY         | 13/21    | upstream reliability is '88.14432989690721'. current run reliability is '61.904761904761905'. drift is 26.23957 and the allowed drift is set to 50. The test should PASS | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                                 |
| DisablingPartitionsTest             | test_disable                                                    | null                                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-779c-440d-973a-8c30a564c6de | FLAKY         | 13/21    | upstream reliability is '86.2547288776797'. current run reliability is '61.904761904761905'. drift is 24.34997 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=DisablingPartitionsTest&test_method=test_disable                                                 |
| RecoveryModeTest                    | test_rolling_restart                                            | null                                                                                                                                  | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-779e-456c-bf98-ee9743de2981 | FLAKY         | 17/21    | upstream reliability is '96.55172413793103'. current run reliability is '80.95238095238095'. drift is 15.59934 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=RecoveryModeTest&test_method=test_rolling_restart                                                |
| WriteCachingFailureInjectionE2ETest | test_crash_all                                                  | {"use_transactions": false}                                                                                                           | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-7795-4310-88bb-b5bf74797973 | FLAKY         | 15/21    | upstream reliability is '98.95287958115183'. current run reliability is '71.42857142857143'. drift is 27.52431 and the allowed drift is set to 50. The test should PASS  | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=WriteCachingFailureInjectionE2ETest&test_method=test_crash_all                                   |
| VerifyConsumerOffsetsThruUpgrades   | test_consumer_group_offsets                                     | {"versions_to_upgrade": 1}                                                                                                            | integration | https://buildkite.com/redpanda/redpanda/builds/72733#01997707-055b-47c7-9272-344d342ff05e | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS                | https://redpanda.metabaseapp.com/dashboard/87-tests?tab=142-dt-individual-test-history&test_class=VerifyConsumerOffsetsThruUpgrades&test_method=test_consumer_group_offsets                        |
</details>
<!--/body-->

michael-redpanda commented:
@c4milo fyi

michael-redpanda commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -137,6 +137,37 @@ Iter validate_range_valid_iterations(
+    std::transform(valid_range_end, end, out_it, [](const type& item) {
+        return generate_error(
+          item,
+          kafka::error_code::unacceptable_credential,
```
change: make this the unauthorized error code

IoannisRP replied:
changed to `cluster_authorization_failed`. I think it's the most fitting from the kafka errors 🤔

michael-redpanda reviewed (commented):
Couple of requests:

1. change the repsonse code to unauthorized user
2. Add a release note as we are slightly changing the behavior of the API.  Something a little more descriptive:

Credentials of superusers will not be permitted to be altered or removed via the Kafka API unless being done by an authenticated superuser.

something like that.  Thanks!

pgellert commented on src/v/kafka/server/handlers/alter_user_scram_credentials.cc:
```diff
@@ -287,6 +318,39 @@ ss::future<response_ptr> alter_user_scram_credentials_handler::handle(
+          klog.trace,
+          "Reauthenticating as Altering user scram credentials for superusers "
+          "requires superuser access");
+        return !ctx.authorized(
+          security::acl_operation::alter,
+          security::default_cluster_name,
+          authz_quiet{false},
+          audit_authz_check::yes,
+          superuser_required::yes);
```
I think here we're missing a check for `ctx.audit()` after the call to `ctx.authorized`. That's to ensure that if auditing fails, we return a `broker_not_available` error code.

I think it would be simpler to do all the authZ checks at the same time. I don't see why we would need two separate AuthZ checks here (and two corresponding audit events).

IoannisRP replied:
Moved away from second authorization to not have to worry about the second auditing event and to simplify the logic.

IoannisRP committed:
k/server: Expose superuser query on the connection_context

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/5e7da97c911a20d7b44e2884a6574eb1f6ec3967..bcd917e68e1f9cbcb6c60e1bb4b918aef4093e7e):
- removed second auth check. Explose util to see if the connection is a superuser, instead
- moved the superuser check prior to deduplication to mirror error handling of normal users

BenPope commented on src/v/kafka/server/tests/alter_user_scram_credentials_test.cc:
```diff
@@ -609,3 +609,242 @@ FIXTURE_TEST(
       resp.data.results[0].error_code, kafka::error_code::duplicate_resource);
 }
+
+FIXTURE_TEST(
```
Should there be a test for when authentication is not enabled on the cluster?

I.e., I think you're missing a condition check; if authentication is not enabled, alter/delete a SCRAM credential for a superuser is not possible.

IoannisRP replied:
while an un-authenticated user should be automatically be treated as a not-superuser, i think you are right that it is better to add the test and be very explicit about it 👍 
I will add one more test.

BenPope replied:
> while an un-authenticated user should be automatically be treated as a not-superuser, i think you are right that it is better to add the test and be very explicit about it 👍 I will add one more test.

Apologies, I meant without authorization:
```
rpk cluster config set kafka_enable_authorization false
```

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/bcd917e68e1f9cbcb6c60e1bb4b918aef4093e7e..70b1ab304f3fee95efc1a9526c27db4ee6aec65d):
- added an extra test case for protecting superusers when auth is turned off

vbotbuildovich commented:
<!--FAILURES SLASH COMMAND https://buildkite.com/redpanda/redpanda/builds/72658-->
<!--9c18d9d0-3fd6-4696-bbbe-4bd5b561f0d1-->
<!--title-->
## Retry command for [Build#72658](https://buildkite.com/redpanda/redpanda/builds/72658)
please wait until all jobs are finished before running the slash command
<!--/title-->
<!--body-->
```
/ci-repeat 1
tests/rptest/tests/partition_force_reconfiguration_test.py::NodeWiseRecoveryTest.test_recovery_local_data_missing@{"wait_for_final_manifest_uploads":true}
```
<!--/body-->

BenPope commented on src/v/kafka/server/tests/alter_user_scram_credentials_test.cc:
```diff
@@ -609,3 +609,329 @@ FIXTURE_TEST(
+    auto deferred_close = ss::defer([&client] { client.stop().get(); });
+    client.connect().get();
+
+    // an user cannot update a superuser even when auth checks are disabled
```
This is AuthZ, right? Please clarify in the comment.

If authorization is not required anywhere in the cluster, why would we not want to be able to update superusers credentials?

It can be done with admin API, right?

IoannisRP replied:
yes, these changes only affect the behavior through the kafka alter_user_scram_credentials api. No changes in the admin api behavior.

BenPope replied:
> yes, these changes only affect the behavior through the kafka alter_user_scram_credentials api. No changes in the admin api behavior.

My point is that when AuthZ is not enabled, we don't usually restrict what can be done.

IoannisRP replied:
that's a fair point. Updated behavior accordingly

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/70b1ab304f3fee95efc1a9526c27db4ee6aec65d..66f31ea2cd9b6ee119498ed844ee262c4cc5f233):
- remove unused var in test
- clarified test description

IoannisRP commented:
changes in [force-push](https://github.com/redpanda-data/redpanda/compare/66f31ea2cd9b6ee119498ed844ee262c4cc5f233..afba28230c45d1f50c1fb1009c15bf01928b8924):
- changed behavior to allow edits to superusers when authz is turned off

vbotbuildovich commented:
/backport v25.2.x
